apis/changelogs/proto/v1alpha1/changelog.proto

```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2024 The Crossplane Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+syntax = "proto3";
+
+import "google/protobuf/struct.proto";
+
+// buf:lint:ignore PACKAGE_DIRECTORY_MATCH
+package changelogs.proto.v1alpha1;
+
+option go_package = "github.com/crossplane/crossplane-runtime/apis/changelogs/proto/v1alpha1";
+
+// ChangeLogService is a service that provides the ability to send change log
+// entries.
+service ChangeLogService {
+  // SendChangeLog sends a change log entry to the change log service.
+  rpc SendChangeLog (SendChangeLogRequest) returns (SendChangeLogResponse) {}
+}
+
+// SendChangeLogRequest represents a single change log entry, with detailed
+// information about the resource that was changed.
+message SendChangeLogRequest {
+  // The name and version of the provider that is making the change to the
+  // resource.
+  string provider = 1;
+
+  // The type of the resource that was changed, e.g. Group, Version, Kind.
+  string type = 2;
+
+  // The name of the resource that was changed.
+  string name = 3;
+
+  // The external name of the resource that was changed.
+  string external_name = 4;
+
+  // The type of operation that was performed on the resource, e.g. Create,
+  // Update, or Delete.
+  OperationType operation = 5;
+
+  // A full snapshot of the resource's state, as observed directly before the
+  // resource was changed.
+  google.protobuf.Struct snapshot = 6;
+
+  // An optional error message that describes any error encountered while
+  // performing the operation on the resource.
+  optional string error_message = 7;
+
+  // An optional additional details that can be provided for further context
+  // about the change.
+  optional google.protobuf.Struct additional_details = 8;
+}
+
+// OperationType represents the type of operation that was performed on a
+// resource.
+enum OperationType {
+  OPERATION_TYPE_UNSPECIFIED = 0;
+  OPERATION_TYPE_CREATE = 1;
+  OPERATION_TYPE_UPDATE = 2;
+  OPERATION_TYPE_DELETE = 3;
+}
+
+// SendChangeLogResponse is the response returned by the ChangeLogService after a
+// change log entry is sent.
+message SendChangeLogResponse {
+  // A boolean indicating whether the change log entry was successfully sent.
+  bool success = 1;
+
+  // A message that provides further information about the result of sending the
+  // change log entry.
+  string message = 2;
```

Comment:
I think gRPC has built in support to return this level of detail as part of the actual error.

https://grpc.io/docs/guides/error/
https://github.com/grpc/grpc-go/blob/e524655/examples/features/error_details/server/main.go#L54

I believe it'd be idiomatic gRPC to have the ChangeLogService just return an empty message if everything worked okay, and return an error if there was one. You could return the [well-known Empty message](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/empty.proto#L43), but probably defining SendChangeLogResponse as an empty message would be more future proof. That way we could add response fields in future if we ever realized we needed them.

Reply:
Great idea, with https://github.com/crossplane/crossplane-runtime/pull/754/commits/afb1c6b1148b69d12412523e3133266c8c810ea2, the response is now empty and any information being returned with the response is through errors.

Server side error handling can be seen in https://github.com/jbw976/change-log-sidecar/blob/main/server/server.go.